app/bundles/LeadBundle/EventListener/CampaignActionAnonymizeUserDataSubscriber.php

```diff
@@ -0,0 +1,144 @@
+        private AuditLogModel $auditLogModel,
+    ) {
+    }
+
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            CampaignEvents::CAMPAIGN_ON_BUILD                  => ['configureAction', 0],
+            LeadEvents::ON_CAMPAIGN_ACTION_ANONYMIZE_USER_DATA => ['anonymizeUserData', 10],
+        ];
+    }
+
+    public function configureAction(CampaignBuilderEvent $event): void
+    {
+        $event->addAction(
+            self::KEY_EVENT_NAME,
+            [
+                'label'                  => 'mautic.lead.lead.events.anonymize',
+                'description'            => 'mautic.lead.lead.events.anonymize_descr',
+                // Kept for BC in case plugins are listening to the shared trigger
+                'eventName'              => LeadEvents::ON_CAMPAIGN_TRIGGER_ACTION, // @phpstan-ignore-line
+                'formType'               => CampaignActionAnonymizeUserDataType::class,
+                'batchEventName'         => LeadEvents::ON_CAMPAIGN_ACTION_ANONYMIZE_USER_DATA,
+            ]
+        );
+    }
+
+    public function anonymizeUserData(PendingEvent $event): void
+    {
+        if (!$event->checkContext(self::KEY_EVENT_NAME)) {
+            return;
+        }
+
+        $properties           = $event->getEvent()->getProperties();
+        $isToPseudonymize     = false;
+        if (array_key_exists('pseudonymize', $properties)) {
+            $isToPseudonymize = (bool) $properties['pseudonymize'];
+        }
+        $leadFilter           = [
+            'force' => [
+                ['column' => 'l.id', 'expr' => 'in', 'value' => $event->getContactIds()],
+            ],
+        ];
+        $leads            = $this->leadModel->getEntities($leadFilter);
+        $companies        = $this->companyModel->getCompaniesByLeads($event->getContactIds());
+        $idFields         = array_merge($properties['fieldsToAnonymize'], $properties['fieldsToDelete']);
+        $fieldFilter      = [
+            'force' => [
+                ['column' => 'f.id', 'expr' => 'in', 'value'  => $idFields],
+            ],
+        ];
+        $fields                       = $this->fieldModel->getRepository()->getEntities($fieldFilter);
+        $deleteFormResultsAndAuditLog = false;
+
+        foreach ($fields as $field) {
+            if (in_array($field->getId(), $properties['fieldsToDelete'])) {
+                [$leads,$companies]           = $this->setDeleteFields($leads, $companies, $field);
+                $deleteFormResultsAndAuditLog = true;
+                continue;
+            }
+
+            if (in_array($field->getId(), $properties['fieldsToAnonymize'])) {
+                [$leads,$companies]           = $this->setHashFields($leads, $companies, $field, $isToPseudonymize);
+                $deleteFormResultsAndAuditLog = true;
+            }
+        }
+
+        if ($deleteFormResultsAndAuditLog) {
```

Comment:
Does this variable mean if we do not have fields to anoynmize or remove we should stop processing campaign action?  Then we need variable  $deleteFormResultsAndAuditLog ?

Reply:
>  I wrote code twice because you didn't have the capacity to understand, and just drop random comments as this one.

Please, can you avoid this? Random comments? WTF?
I understand you code, but it's  over-engineered.  You didn't do relevant answer to any of my question.

Reply:
First, "hell", "WTF"...